### PR TITLE
task : fix bitails false positive responses in broadcasting

### LINKED ISSUE #582: [Task] Resolve issue with Bitails returning false positive responses in broadcasting

### PR DESCRIPTION
**Pull request by cain4chain**
## Description of Changes

- Add double spend error code handling to bitails broadcast response methods
- Add descriptive error field for ECONNRESET and ECONNREFUSED errors
- Add corresponding test cases
- Handle response type conflicts for "code" field (`int `or `string`)

Example Response
`Bitails response Status: 201
Bitails response Body: [{"txid":"411193c81c7effbb42a5583cd0aacd5ef02daa94174491279d87c99410dc0bad"},{"txid":"c79082eee4433ea7bbea40abca8aeed0526d07e569e73b114accce04da3209fd"},{"txid":"0f4b4f89590b9296f72cf60aae70892372f2eaf66275c4912803715cc3c3c89d"},{"txid":"342c524c6dc291df790e1cdabae1d95f7d6a82229446d11a772e1817c192f766"},{"txid":"f9b0aadcb2564f028d2e216d0af6da4ceb348aa990cb4c43dedd57f04755b0a2"},{"error":{"code":-26,"message":"double-spend-attempt"}},{"error":{"code":"ECONNRESET"}},{"error":{"errno":-111,"code":"ECONNREFUSED","syscall":"connect","address":"127.0.0.1","port":1113}}]`

## Linked Issues / Tickets

Closes #582

Describe the tests you've added or any testing steps you've taken.

- [ ] I have added new unit tests
- [ ] All tests pass locally
- [ ] I have tested manually in my local environment

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have run the linter

### CODE REVIEW

File: pkg/services/internal/bitails/broadcast.go [hunk 1]
```diff
@@ -105,13 +106,27 @@ func (b *Bitails) classifyResponseError(resp broadcastResponse, result *wdk.Post
...
+	case ErrorCodeMissingInputs:
+		result.Result = wdk.PostedTxIDResultMissingInputs
+		result.DoubleSpend = true
+		shouldReturnError = true
 	default:
 		result.Result = wdk.PostedTxIDResultError
 		result.Error = fmt.Errorf("broadcast error code %d: %s", resp.Error.Code, msg)
+		upperMsg := strings.ToUpper(msg)
+		if strings.Contains(upperMsg, ErrorTokenECONNRESET) || strings.Contains(upperMsg, ErrorTokenECONNREFUSED) {
+			if strings.Contains(upperMsg, ErrorTokenECONNREFUSED) {
```
Comment (chris-4chain) [inline pkg/services/internal/bitails/broadcast.go]: I'm not sure if this kind of errors should be handled here. It looks like some network problem - and I though it is handled by `resty` and returned (from resty) as an error

File: pkg/services/internal/bitails/broadcast.go [hunk 2]
```diff
@@ -105,13 +106,27 @@ func (b *Bitails) classifyResponseError(resp broadcastResponse, result *wdk.Post
...
 		result.Result = wdk.PostedTxIDResultAlreadyKnown
 		result.AlreadyKnown = true
-	case ErrorCodeMissingInputs:
+	case ErrorCodeDoubleSpend:
 		result.Result = wdk.PostedTxIDResultDoubleSpend
 		result.DoubleSpend = true
 		shouldReturnError = true
+	case ErrorCodeMissingInputs:
+		result.Result = wdk.PostedTxIDResultMissingInputs
+		result.DoubleSpend = true
```
Comment (chris-4chain) [inline pkg/services/internal/bitails/broadcast.go]: Why is this still marked as Double spend?

File: pkg/services/internal/bitails/broadcast.go (hunk 1 above)
Comment (cain4chain) [reply pkg/services/internal/bitails/broadcast.go]: Yes, found the issue and is updated now. There was a alternative type issue in the response from `int` to `string`. 
Now we still use the conditional logic but we have the correct value for processing. Let me know what you think of the changes, thanks!

File: pkg/storage/internal/actions/process.go [hunk 3]
```diff
@@ -369,7 +369,16 @@ func (p *process) broadcastTxs(ctx context.Context, txIDs []string, isDelayed bo
 
 	logger.DebugContext(ctx, "Categorizing transactions by status")
 
-	for txID, currentStatus := range knownTxStatusesLookup {
+	for _, txID := range txIDs {
+		currentStatus, ok := knownTxStatusesLookup[txID]
```
Comment (chris-4chain) [inline pkg/storage/internal/actions/process.go:373]: Many thanks for finding and fixing a place where we wrongly changed the order of provided `txIDs`.

I think, we don't have to worry about `!ok` case - because it should not happen. Check how `getKnownTxStatuses` works.

So, instead of `DebugLog` and proceeding I would return an error.

File: pkg/services/internal/bitails/broadcast.go (hunk 1 above)
Comment (dorzepowski) [reply pkg/services/internal/bitails/broadcast.go]: I see approaches that I would prefer
Option 1:
1. in broadcastError UnmarshalJSON  assign custom error code if the result is one of  those
2. add `case` statement for those new error codes

Option 2:
1. in broadcastError UnmarshalJSON  unmarshal always to some string code (which also could be an enumerated)
2. use in `case` statements those enumerated strings

File: pkg/services/internal/bitails/broadcast.go [hunk 4]
```diff
@@ -105,13 +133,27 @@ func (b *Bitails) classifyResponseError(resp broadcastResponse, result *wdk.Post
...
+		// result.DoubleSpend = true
+		shouldReturnError = true
+	case ErrorCodeMissingInputs:
+		result.Result = wdk.PostedTxIDResultMissingInputs
 		shouldReturnError = true
 	default:
 		result.Result = wdk.PostedTxIDResultError
 		result.Error = fmt.Errorf("broadcast error code %d: %s", resp.Error.Code, msg)
+		upperMsg := strings.ToUpper(msg)
+		if strings.Contains(upperMsg, ErrorTokenECONNRESET) || strings.Contains(upperMsg, ErrorTokenECONNREFUSED) {
```
Comment (dorzepowski) [inline pkg/services/internal/bitails/broadcast.go]: It would be good, to add example in tests for those cases

**Review by chris-4chain — APPROVED** [review]

File: pkg/services/internal/bitails/broadcast.go (hunk 2 above)
Comment (cain4chain) [reply pkg/services/internal/bitails/broadcast.go]: This is how it was setup on TS- wallet-toolbox for missing inputs. But since we now define missing inputs error in our code I think its safe to remove this. 

I will do some more manual testing for missing inputs though in another task just to cover some edge cases.

**Review by mgosek-4chain — APPROVED** [review]

**Review by jakubmkowalski — APPROVED** [review]
